src/transformStateWithClones.js

```diff
@@ -7,7 +7,40 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let copyState = { ...state };
+  const previousStateVers = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties': {
+        copyState = Object.assign(copyState, action.extraData);
+        previousStateVers.push({ ...copyState });
+        break;
+      }
+
+      case 'removeProperties': {
+        for (const keyToRemove of action.keysToRemove) {
+          delete copyState[keyToRemove];
+        }
+        previousStateVers.push({ ...copyState });
+        break;
+      }
+
+      case `clear`: {
+        for (const key in copyState) {
+          delete copyState[key];
+        }
+        previousStateVers.push({ ...copyState });
+        break;
+      }
+
+      default: {
+        throw new Error('Unknown action');
```

Comment:
Need to add more error description. You can add an `action.type` that falls into the default state.